File: psana_examples/trunk/src/DumpTimepix.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpTimepix...
//
// Author List:
//      Andy Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpTimepix.h"

//-----------------
// C/C++ Headers --
//-----------------

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
// to work with detector data include corresponding 
// header from psddl_psana package
#include "psddl_psana/timepix.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

// This declares this class as psana module
using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpTimepix)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpTimepix::DumpTimepix (const std::string& name)
  : Module(name)
  , m_src()
{
  // get the values from configuration or use defaults
  m_src = configStr("source", "DetInfo(:Timepix)");
}

//--------------
// Destructor --
//--------------
DumpTimepix::~DumpTimepix ()
{
}

/// Method which is called at the beginning of the calibration cycle
void 
DumpTimepix::beginCalibCycle(Event& evt, Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::Timepix::ConfigV1> config = env.configStore().get(m_src);
  if (config.get()) {

    WithMsgLog(name(), info, str) {
      str << "Timepix::ConfigV1:";

      str << "\n  readoutSpeed = " << int(config->readoutSpeed());
      str << "\n  triggerMode = " << int(config->triggerMode());
      str << "\n  shutterTimeout = " << config->shutterTimeout();
      str << "\n  dac0Ikrum = " << config->dac0Ikrum();
      str << "\n  dac0Disc = " << config->dac0Disc();
      str << "\n  dac0Preamp = " << config->dac0Preamp();
      str << "\n  dac0BufAnalogA = " << config->dac0BufAnalogA();
      str << "\n  dac0BufAnalogB = " << config->dac0BufAnalogB();
      str << "\n  dac0Hist = " << config->dac0Hist();
      str << "\n  dac0ThlFine = " << config->dac0ThlFine();
      str << "\n  dac0ThlCourse = " << config->dac0ThlCourse();
      str << "\n  dac0Vcas = " << config->dac0Vcas();
      str << "\n  dac0Fbk = " << config->dac0Fbk();
      str << "\n  dac0Gnd = " << config->dac0Gnd();
      str << "\n  dac0Ths = " << config->dac0Ths();
      str << "\n  dac0BiasLvds = " << config->dac0BiasLvds();
      str << "\n  dac0RefLvds = " << config->dac0RefLvds();
      str << "\n  dac1Ikrum = " << config->dac1Ikrum();
      str << "\n  dac1Disc = " << config->dac1Disc();
      str << "\n  dac1Preamp = " << config->dac1Preamp();
      str << "\n  dac1BufAnalogA = " << config->dac1BufAnalogA();
      str << "\n  dac1BufAnalogB = " << config->dac1BufAnalogB();
      str << "\n  dac1Hist = " << config->dac1Hist();
      str << "\n  dac1ThlFine = " << config->dac1ThlFine();
      str << "\n  dac1ThlCourse = " << config->dac1ThlCourse();
      str << "\n  dac1Vcas = " << config->dac1Vcas();
      str << "\n  dac1Fbk = " << config->dac1Fbk();
      str << "\n  dac1Gnd = " << config->dac1Gnd();
      str << "\n  dac1Ths = " << config->dac1Ths();
      str << "\n  dac1BiasLvds = " << config->dac1BiasLvds();
      str << "\n  dac1RefLvds = " << config->dac1RefLvds();
      str << "\n  dac2Ikrum = " << config->dac2Ikrum();
      str << "\n  dac2Disc = " << config->dac2Disc();
      str << "\n  dac2Preamp = " << config->dac2Preamp();
      str << "\n  dac2BufAnalogA = " << config->dac2BufAnalogA();
      str << "\n  dac2BufAnalogB = " << config->dac2BufAnalogB();
      str << "\n  dac2Hist = " << config->dac2Hist();
      str << "\n  dac2ThlFine = " << config->dac2ThlFine();
      str << "\n  dac2ThlCourse = " << config->dac2ThlCourse();
      str << "\n  dac2Vcas = " << config->dac2Vcas();
      str << "\n  dac2Fbk = " << config->dac2Fbk();
      str << "\n  dac2Gnd = " << config->dac2Gnd();
      str << "\n  dac2Ths = " << config->dac2Ths();
      str << "\n  dac2BiasLvds = " << config->dac2BiasLvds();
      str << "\n  dac2RefLvds = " << config->dac2RefLvds();
      str << "\n  dac3Ikrum = " << config->dac3Ikrum();
      str << "\n  dac3Disc = " << config->dac3Disc();
      str << "\n  dac3Preamp = " << config->dac3Preamp();
      str << "\n  dac3BufAnalogA = " << config->dac3BufAnalogA();
      str << "\n  dac3BufAnalogB = " << config->dac3BufAnalogB();
      str << "\n  dac3Hist = " << config->dac3Hist();
      str << "\n  dac3ThlFine = " << config->dac3ThlFine();
      str << "\n  dac3ThlCourse = " << config->dac3ThlCourse();
      str << "\n  dac3Vcas = " << config->dac3Vcas();
      str << "\n  dac3Fbk = " << config->dac3Fbk();
      str << "\n  dac3Gnd = " << config->dac3Gnd();
      str << "\n  dac3Ths = " << config->dac3Ths();
      str << "\n  dac3BiasLvds = " << config->dac3BiasLvds();
      str << "\n  dac3RefLvds = " << config->dac3RefLvds();
    }

  }
}

/// Method which is called with event data, this is the only required 
/// method, all other methods are optional
void 
DumpTimepix::event(Event& evt, Env& env)
{
  shared_ptr<Psana::Timepix::DataV1> data1 = evt.get(m_src);
  if (data1.get()) {
    WithMsgLog(name(), info, str) {
      str << "Timepix::DataV1:";

      str << "\n  timestamp = " << data1->timestamp();
      str << "\n  frameCounter = " << data1->frameCounter();
      str << "\n  lostRows = " << data1->lostRows();

      const ndarray<uint16_t, 2>& img = data1->data();
      str << "\n  data =";
      str << " (" << img.shape()[0] << ", " << img.shape()[0] << ")";
      for (int i = 0; i < 10; ++ i) {
        str << " " << img[0][i];
      }
      str << " ...";
    }
  }

  shared_ptr<Psana::Timepix::DataV2> data2 = evt.get(m_src);
  if (data2.get()) {
    WithMsgLog(name(), info, str) {
      str << "Timepix::DataV2:";

      str << "\n  timestamp = " << data2->timestamp();
      str << "\n  frameCounter = " << data2->frameCounter();
      str << "\n  lostRows = " << data2->lostRows();

      const ndarray<uint16_t, 2>& img = data2->data();
      str << "\n  data =";
      str << " (" << img.shape()[0] << ", " << img.shape()[0] << ")";
      for (int i = 0; i < 10; ++ i) {
        str << " " << img[0][i];
      }
      str << " ...";
    }
  }
}
  

} // namespace psana_examples

File: python_examples/trunk/src/dump_timepix.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id: dump_timepix.py 3076 2012-03-14 21:51:06Z salnikov@SLAC.STANFORD.EDU $
#
# Description:
#  Pyana user analysis module dump_timepix...
#
#------------------------------------------------------------------------

"""User analysis module for pyana framework.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id: dump_timepix.py 3076 2012-03-14 21:51:06Z salnikov@SLAC.STANFORD.EDU $

@author Andy Salnikov
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision: 3076 $"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

#---------------------
#  Class definition --
#---------------------
class dump_timepix (object) :
    """example analysis module which dumps Timepix objects"""

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        """Class constructor takes the name of the data source.

        @param source   data source
        """
        
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------

    def beginjob( self, evt, env ) :
        
        config = env.getConfig(xtc.TypeId.Type.Id_TimepixConfig, self.m_src)
        if config:
        
            print "%s: %s" % (config.__class__.__name__, self.m_src)
            
            print "  readoutSpeed =", config.readoutSpeed()
            print "  triggerMode =", config.triggerMode()
            try:
                # ConfigV1
                print "  shutterTimeout =", config.shutterTimeout()
            except:
                pass
            try:
                # ConfigV2
                print "  timepixSpeed =", config.timepixSpeed()
            except:
                pass
            print "  dac0Ikrum =", config.dac0Ikrum()
            print "  dac0Disc =", config.dac0Disc()
            print "  dac0Preamp =", config.dac0Preamp()
            print "  dac0BufAnalogA =", config.dac0BufAnalogA()
            print "  dac0BufAnalogB =", config.dac0BufAnalogB()
            print "  dac0Hist =", config.dac0Hist()
            print "  dac0ThlFine =", config.dac0ThlFine()
            print "  dac0ThlCourse =", config.dac0ThlCourse()
            print "  dac0Vcas =", config.dac0Vcas()
            print "  dac0Fbk =", config.dac0Fbk()
            print "  dac0Gnd =", config.dac0Gnd()
            print "  dac0Ths =", config.dac0Ths()
            print "  dac0BiasLvds =", config.dac0BiasLvds()
            print "  dac0RefLvds =", config.dac0RefLvds()
            print "  dac1Ikrum =", config.dac1Ikrum()
            print "  dac1Disc =", config.dac1Disc()
            print "  dac1Preamp =", config.dac1Preamp()
            print "  dac1BufAnalogA =", config.dac1BufAnalogA()
            print "  dac1BufAnalogB =", config.dac1BufAnalogB()
            print "  dac1Hist =", config.dac1Hist()
            print "  dac1ThlFine =", config.dac1ThlFine()
            print "  dac1ThlCourse =", config.dac1ThlCourse()
            print "  dac1Vcas =", config.dac1Vcas()
            print "  dac1Fbk =", config.dac1Fbk()
            print "  dac1Gnd =", config.dac1Gnd()
            print "  dac1Ths =", config.dac1Ths()
            print "  dac1BiasLvds =", config.dac1BiasLvds()
            print "  dac1RefLvds =", config.dac1RefLvds()
            print "  dac2Ikrum =", config.dac2Ikrum()
            print "  dac2Disc =", config.dac2Disc()
            print "  dac2Preamp =", config.dac2Preamp()
            print "  dac2BufAnalogA =", config.dac2BufAnalogA()
            print "  dac2BufAnalogB =", config.dac2BufAnalogB()
            print "  dac2Hist =", config.dac2Hist()
            print "  dac2ThlFine =", config.dac2ThlFine()
            print "  dac2ThlCourse =", config.dac2ThlCourse()
            print "  dac2Vcas =", config.dac2Vcas()
            print "  dac2Fbk =", config.dac2Fbk()
            print "  dac2Gnd =", config.dac2Gnd()
            print "  dac2Ths =", config.dac2Ths()
            print "  dac2BiasLvds =", config.dac2BiasLvds()
            print "  dac2RefLvds =", config.dac2RefLvds()
            print "  dac3Ikrum =", config.dac3Ikrum()
            print "  dac3Disc =", config.dac3Disc()
            print "  dac3Preamp =", config.dac3Preamp()
            print "  dac3BufAnalogA =", config.dac3BufAnalogA()
            print "  dac3BufAnalogB =", config.dac3BufAnalogB()
            print "  dac3Hist =", config.dac3Hist()
            print "  dac3ThlFine =", config.dac3ThlFine()
            print "  dac3ThlCourse =", config.dac3ThlCourse()
            print "  dac3Vcas =", config.dac3Vcas()
            print "  dac3Fbk =", config.dac3Fbk()
            print "  dac3Gnd =", config.dac3Gnd()
            print "  dac3Ths =", config.dac3Ths()
            print "  dac3BiasLvds =", config.dac3BiasLvds()
            print "  dac3RefLvds =", config.dac3RefLvds()
            
            try:
                # ConfigV2
                print "  chipCount =", config.chipCount()
                print "  driverVersion =", config.driverVersion()
                print "  firmwareVersion =", config.firmwareVersion()
                print "  pixelThreshSize =", config.pixelThreshSize()
                print "  pixelThresh =", config.pixelThresh()
                print "  chip names =", config.chip0Name(), config.chip1Name(), config.chip2Name(), config.chip3Name()
                print "  chip IDs =", config.chip0ID(), config.chip1ID(), config.chip2ID(), config.chip3ID()
            except:
                pass


    def event( self, evt, env ) :

        data = evt.get(xtc.TypeId.Type.Id_TimepixData, self.m_src)
        if data:
            print "%s: %s" % (data.__class__.__name__, self.m_src)
            
            print "  timestamp =", data.timestamp()
            print "  frameCounter =", data.frameCounter()
            print "  lostRows =", data.lostRows()
            
            img = data.data()
            print "  data.shape =", img.shape
            print "  data =", img

    def endjob( self, evt, env ) :
        
        pass
